[DOC] Add how-to guides for local/HPC parallelization (#713)
nipoppy/nipoppy: docs/source/how_to_guides/parallelization/hpc_scheduler.md

# Parallelizing pipeline runs on HPC systems

> This guide shows how to parallelize pipeline runs on {term}`HPC` systems that use job schedulers supported by the Nipoppy tools.

Reviewer: ```suggestion
This guide shows how to parallelize pipeline runs on {term}`HPC` systems that use job schedulers supported by Nipoppy.
```

> Currently, we have built-in support for the [Slurm](https://slurm.schedmd.com/overview.html) and [SGE](https://en.wikipedia.org/wiki/Oracle_Grid_Engine) job schedulers only.

Reviewer: ```suggestion
Currently, we have built-in support for the [Slurm](https://slurm.schedmd.com/overview.html) and [SGE](https://en.wikipedia.org/wiki/Oracle_Grid_Engine) job schedulers.
```

> However, it may also be possible to use other job schedulers [with some manual work](#support-for-other-job-schedulers).

Reviewer: ```suggestion
However, it is possible to [manually add another job scheduler](#support-for-other-job-schedulers).
```

```{important}
Although the default template job script is designed to work with minimal user configuration, each {term}`HPC` system is different, and some may require different/additional parameters to be set. See the [Further customization](#further-customization) section for how deeper configuration can be achieved.

If the default Slurm/SGE configurations do not work for you, please consider [opening an issue on our GitHub repository](https://github.com/nipoppy/nipoppy/issues/new/choose) so that we can improve our HPC support.
```

## Configuring main HPC options

### Global settings

The default global configuration file has two {term}`HPC`-related fields that should be updated as needed:

> ```{literalinclude} ../../../../nipoppy/data/examples/sample_global_config.json

Reviewer: We could include the preamble for sourcing the venv/conda in the sample global config.

Author: The issue is that paths/env names would be different, and they would have to delete either the line for the env method they're not using, and JSON doesn't allow comments...

But not a bad idea

`HPC_PREAMBLE` is a list of Bash commands that should executed at the beginning of **every** job.
Importantly, there should be a command for activating the Nipoppy Python environment.
* For {term}`venv` environments, the command would be something like this: `"source <PATH_TO_NIPOPPY_VENV>/bin/activate"`
* For {term}`conda` environments, the command would instead be something like this: `"source ~/.bashrc; conda activate <NIPOPPY_ENV_NAME>"`

#### `[[HPC_ACCOUNT_NAME]]`

The value for the `[[HPC_ACCOUNT_NAME]]` field in the `SUBSTITUTIONS` dictionary should be set to the account name/ID the job will be associated with.
By default this will be passed as `--account-name` in Slurm systems and `-q` in SGE systems during job submission.
This can be left blank if these options are not needed.

```{attention}
If your HPC system needs flags other than `--account-name` or `-q` need to be set, you will have to modify the template job submission script: see the [Further customization](#further-customization) section for more information.
```

### Pipeline-specific settings

Job time limit and CPU and memory requests can be configured separately for each pipeline via the HPC config file.

> Look for this file inside the pipeline config directory at {{dpath_pipelines}}`/{bidsification,processing,extraction}/<PIPELINE_NAME>/<PIPELINE_VERSION>` -- it is most likely called `hpc.json` or `hpc_config.json` and should look something like this:

Reviewer: We can specify the field to look for in the pipeline `config.json` file, in case it is name differently.

* `TIME`: time limit. Passed as `--time` in Slurm jobs and `-l h_rt` in SGE jobs.
* `CORES`: number of CPUs requested. Passed as `--cpus-per-task` in Slurm jobs and ignored in SGE jobs.
* `MEMORY`: amount of memory requested. Passed as `--mem` in Slurm jobs and `-l h_vmem` in SGE jobs.
* `ARRAY_CONCURRENCY_LIMIT`: maximum number of jobs in the array that can be run at the same time. Set as part of `--array` specification in Slurm jobs and passed as `--tc` in SGE jobs.

## Submitting HPC jobs via `nipoppy` commands

To run a pipeline on an HPC, use the `--hpc` option to specify the HPC job scheduler when running the [`nipoppy bidsify`](<project:../../cli_reference/bidsify.rst>), [`nipoppy process`](<project:../../cli_reference/process.rst>), or [`nipoppy extract`](<project:../../cli_reference/extract.rst>) commands:

```console
$ nipoppy <SUBCOMMAND> \
    --dataset <NIPOPPY_PROJECT_ROOT> \
    --pipeline <PIPELINE_NAME> \
    --hpc slurm
    # other desired options
    # ...
```

This will submit a job array (one job per participant/session to run) through the requested job scheduler.

> Currently, only `'slurm'` and `'sge'` have built-in support, but it is possible to [add other cluster types](#support-for-other-job-schedulers).

Reviewer: ```suggestion
Currently, only `'slurm'` and `'sge'` have built-in support, but it is possible to [add another job scheduler](#support-for-other-job-schedulers).
```
